Adds CSV support to the LearnerListView

pomegranited opened the pull request:
This change allows the LearnerListView API call to return paginated CSV data, allowing it to be used to provide a Learner Profile Report for a given course.  [Example CSV attached](https://github.com/edx/edx-analytics-data-api/files/433894/OpenCraft-Analytics101x-now--2016-08-11T10-25-16.00-00--learners.csv.txt).

* Pagination links are set in the response header, following the [convention used by the github API](https://developer.github.com/guides/traversing-with-pagination/).
* Adds a 'fields' query string parameter, which can be used to limit (and in the case of a CSV response, sort) the list of fields returned.

**Discussions**: Architecture discussed extensively with @mulby and the analytics team.

**Sandbox URL**: TBD - sandbox is being provisioned.

**Partner information**: 3rd party-hosted open edX instance

**Testing instructions**:
<details>
<summary>Testing setup is extensive, sorry!</summary>

1. Set up analyicstack according to [edx instructions](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/installation/analytics/install_analytics.html).
1. Create LMS user and tracking data:
    1. [Run Studio](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/installation/analytics/start_analytics.html#starting-the-open-edx-lms)
    1. Create and publish a course in Studio, with one or more Problem Units.
    1. [Run LMS](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/installation/analytics/start_analytics.html#starting-the-open-edx-lms)
    1. Register 2 or 3 student accounts, and enrol them in your course.
    1. Have them attempt some problems.  To test all the different data permutations:
        * perform multiple attempts
        * make some students get the problems correct, and some incorrect.
    1. Check that `/edx/var/log/tracking/tracking.log` is receiving these events.
    1. Check that hdfs is receiving the updated tracking logs (cron'd every 1 min),
       by browsing to the [hdfs explorer](http://localhost:50070/explorer.html), and navigating to the `data` folder.
1. Set up analytics ElasticSearch.  

    *NB* The analyticstack already runs ES v0.90 for the LMS/Forum, but the analytics API and pipeline require ES v1.5.   There are [other possible approaches](https://openedx.atlassian.net/wiki/display/AN/Learner+Analytics#LearnerAnalytics-ElasticSearch), but here I decided to run two ES instances on vagrant, on different ports, using the [Analytics Data API instructions](https://github.com/edx/edx-analytics-data-api):
       
    ``` 
vagrant ssh
sudo su analytics_api
source /edx/app/analytics_api/venvs/analytics_api/bin/activate
# Runs ES on port 9223 by default.
~/analytics_api$ make test.install_elasticsearch
~/analytics_api$ make test.run_elasticsearch
    ```

1. Install the analytics pipeline on your vagrant instance, by running a `remote-task`:
    1. On your local machine, ensure that the `edx-analytics-pipeline` repo is created and installed as per [step 3.6.11. of setting up the analyticstack](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/installation/analytics/install_analytics.html#id7).
    1. Go to the `edx-analytics-pipeline` repo dir, and activate the virtualenv.
    1. Provide an `override.cfg` configuration file to use when running tasks.
       
        Here, we point the analytics pipeline to the ES instance created in step 3,
        and allow empty module-engagement data inserts, to account for our sparse tracking data.

        ``` 
cp config/devstack.cfg override.cfg
vim override.cfg # make changes below
diff -u config/devstack.cfg  override.cfg
--- config/devstack.cfg 2016-07-27 09:30:36.422732328 +0000
+++ override.cfg    2016-07-27 11:54:26.860951553 +0000
@@ -78,9 +78,10 @@  
 dropoff_threshold = 0.05
 
 [elasticsearch]
-# Point to the vagrant host's port 9201 where we assume elasticsearch is running
-host = http://172.17.0.1:9201/
+# Point to the analytics data API's test elasticsearch cluster on port 9233
+host = http://127.0.0.1:9223/
 
 [module-engagement]
 alias = roster
 number_of_shards = 5
+# Allow empty inserts, since our tracking data is sparse
+allow_empty_insert = true
        ```

    1. Run the `remote-task` `ImportEnrollmentsIntoMysql`, to import enrollment records required by the module engagement task.

      This will install the chosen repo:branch of the analytics pipeline on your vagrant instance to `/var/lib/analytics-tasks/analyticstack/repo`
      before running the actual task.
      
      Ensure that the `--interval` parameter is set to cover the days present in your tracking log.

      Ensure that `$PIPELINE_REPO` and `$PIPELINE_BRANCH` reflect the repo and branch of the edx-analytics-pipeline you wish to install.

        ```
export WHEEL_URL=http://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
VAGRANT_PATH='/path/to/your/Vagrantfile'
PIPELINE_REPO='https://github.com/open-craft/edx-analytics-pipeline.git'
PIPELINE_BRANCH="jill/module-engagement-add-fields"
TODAY=`date +%Y-%m-%d`
TOMORROW=`date --date="tomorrow" +%Y-%m-%d`
remote-task --wheel-url $WHEEL_URL \
    --vagrant-path $VAGRANT_PATH   \
    --repo "$PIPELINE_REPO" \
    --branch "$PIPELINE_BRANCH" \
    --remote-name analyticstack  \
    --override-config "`pwd`/override.cfg"  \
    --wait \
 
 ImportEnrollmentsIntoMysql --interval 2016-07-01-$TOMORROW --local-scheduler --n-reduce-tasks 2 --remote-log-level DEBUG
        ```

   1. Run the `ModuleEngagementWorkflowTask` to import learner profile and engagement data into ElasticSearch.

       *NB* Once you've successfully installed the analytics pipeline, you can run subsequent analytics pipeline tasks directly on your vagrant VM with this much briefer process.

        ```
vagrant ssh
sudo su hadoop
cd /var/lib/analytics-tasks/analyticstack/repo
source /var/lib/analytics-tasks/analyticstack/venv/bin/activate
TODAY=`date +%Y-%m-%d`
launch-task ModuleEngagementWorkflowTask --date $TODAY --local-scheduler --n-reduce-tasks 2 --remote-log-level DEBUG
        ```

1. Run the Analytics Data API:
    1. Checkout the branch for this PR into `analytics_api`.
    1. Update pip requirements (one dependency has changed):
       
        ```
vagrant ssh
source /edx/app/analytics_api/venvs/analytics_api/bin/activate
cd /edx/app/analytics_api/analytics_api
make requirements
        ```

    1. Adjust the settings to point to our ES instance and index/alias:

        ```
vagrant ssh
sudo cp /edx/etc/analytics_api.yml /edx/etc/analytics_api.yml.orig
sudo vim /edx/etc/analytics_api.yml  # apply changes below
diff -u analytics_api.yml.orig analytics_api.yml
--- analytics_api.yml.orig  2016-07-27 12:00:23.363453820 +0000
+++ analytics_api.yml   2016-07-27 12:00:31.278970508 +0000
@@ -24,8 +24,8 @@
 ELASTICSEARCH_AWS_SECRET_ACCESS_KEY: null
 ELASTICSEARCH_CONNECTION_CLASS: null
 ELASTICSEARCH_CONNECTION_DEFAULT_REGION: us-east-1
-ELASTICSEARCH_LEARNERS_HOST: localhost
-ELASTICSEARCH_LEARNERS_INDEX: roster_1_2
+ELASTICSEARCH_LEARNERS_HOST: localhost:9223    # matches ES host:port from step N
+ELASTICSEARCH_LEARNERS_INDEX: roster           # matches alias of roster index used by pipeline.
 ELASTICSEARCH_LEARNERS_UPDATE_INDEX: index_updates
 EMAIL_HOST: localhost
 EMAIL_HOST_PASSWORD: mail_password
        ```

   1. [Create an API key for an existing LMS staff user](http://edx-data-analytics-api.readthedocs.io/en/latest/authentication.html#edx-data-analytics-api-authentication), e.g.

        ```
sudo su analytics_api
source /edx/app/analytics_api/venvs/analytics_api/bin/activate
cd /edx/app/analytics_api/analytics_api
./manage.py set_api_key staff token  # 'token' becomes an API key
        ```

    1. [Run the Analytics Data API server](http://edx-data-analytics-api.readthedocs.io/en/latest/setup.html#run-the-server):

        ```
./manage.py runserver 0.0.0.0:8100 --insecure
        ```

1. View the analytics pipeline data via the Analytics API:
    1. Using the swagger.io UI:
       
        *NB* We need to upgrade swagger.ui to allow the configuration of more Response Content Type options, so the only response type option currently available via this UI is `application/json`.
        1. Visit [the swagger.io UI](http://localhost:8100)
        1. Enter your token in the "API Key" field, and click Explore
        1. Try out GET /authenticated request to ensure 200 response returned
        1. Try out GET /api/v0/learners/ request, with your `course_id` filled in.
    1. Using `curl`:

        ```
 API_TOKEN="<your API token>"
 COURSE_ID="<your URL-encoded course_id>"
 
 # View JSON description of the API call (note "text/csv" in the "renders" list):
 curl -X OPTIONS \
    --header "Authorization: Token $API_TOKEN" \
    "http://localhost:8100/api/v0/learners/?course_id=$COURSE_ID"
 
 # View CSV learner profile data, including response headers:
curl -X GET --include
    --header "Authorization: Token $API_TOKEN" \
    --header "Accept: text/csv" \
    "http://localhost:8100/api/v0/learners/?course_id=$COURSE_ID"
 
 # View paginated CSV learner profile data (e.g. one learner per page), including response headers:
# Note the "Link" header, which is added if there is >1 page of data.
curl -X GET --include \
    --header "Authorization: Token $API_TOKEN" \
    --header "Accept: text/csv" \
    "http://localhost:8100/api/v0/learners/?course_id=$COURSE_ID&page_size=1"
        ```

</details>

**Troubleshooting**:

<details>
<summary>Troubleshooting your test installation is also involved.</summary>

1. Check the output of `remote-task` and/or `launch-task` for errors or warnings.
1. Check your `/var/lib/analytics-tasks/analyticstack/repo/override.cfg` to ensure it contains the configuration options you require.
1. See the [analytics pipeline readthedocs](http://edx-analytics-pipeline-reference.readthedocs.io/en/latest/workflow_entry_point.html) for details on `launch-task` parameters and configuration.
1. Check if the data made it into in mysql:

    ```
vagrant@precise64:~$ mysql -u root reports
mysql> show tables;
+-----------------------------------------+
| Tables_in_reports                       |
+-----------------------------------------+
| course_enrollment_birth_year_daily      |
| course_enrollment_daily                 |
| course_enrollment_education_level_daily |
| course_enrollment_gender_daily          |
| course_enrollment_mode_daily            |
| module_engagement                       |
| module_engagement_metric_ranges         |
| table_updates                           |
+-----------------------------------------+
8 rows in set (0.00 sec)
mysql> select * from course_enrollment_daily;  # should show your students' enrollments
mysql> select * from module_engagement;        # should show your students' problem attempts
    ```

1. Check the alias was added to ElasticSearch:
    1. Add the ES port to your Vagrantfile, and run `vagrant reload`

   ```
  if not ENV['VAGRANT_NO_PORTS']
    ...
    config.vm.network :forwarded_port, guest: 9223, host: 9223  # Elasticsearch v1.5
    ....
```
    * View the [list of ElasticSearch index aliases](http://localhost:9223/_aliases): should contain
      one called `roster`
    * View the [roster alias field mapping](http://localhost:9223/roster/_mapping): should contain a `roster_entry` mapping, with fields like `course_id` and `username`.

</details>

**Reviewers**
- [x] @haikuginger 
- [x] @thallada

edx-webhook commented:
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-analytics-data-api&number=128

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-1360](https://openedx.atlassian.net/browse/OSPR-1360) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-analytics-data-api/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

haikuginger commented on analytics_data_api/v0/serializers.py:
```diff
@@ -45,6 +46,40 @@ class ModelSerializerWithCreatedField(serializers.ModelSerializer):
+        request = self.context.get('request')
+        if request:
+            request_fields = request.QUERY_PARAMS.get('fields')
+            if request_fields:
```
Can we do an explicit `if request_fields is not None` here?

pomegranited replied:
Ah yes.. I have a bad habit of relying on language-specific `False`.  Thanks for catching that!

Addressed with [f5b2ce2](https://github.com/edx/edx-analytics-data-api/pull/128/commits/f5b2ce2b7da0aeb684f110699d7cbb4866f2ea3e).

haikuginger commented on analytics_data_api/v0/views/learners.py:
```diff
@@ -2,12 +2,9 @@
 """
 import logging
-
+from django.conf import settings
```
Nit: Can we keep the newline between stdlib imports and Django imports?

pomegranited replied:
Addressed with [63979d2](https://github.com/edx/edx-analytics-data-api/pull/128/commits/63979d2453205f4e6abee1bdd7c30e1828b9015b).

haikuginger commented on analytics_data_api/v0/views/learners.py:
```diff
@@ -124,16 +121,24 @@ class LearnerListView(LastUpdateMixin, CourseViewMixin, generics.ListAPIView):
+        Pagination links are also returned in the response's header, e.g.
+            Link: <next_url>; rel="next", <previous_url>; rel="prev";
+
+        If the returned object is CSV, only the 'results' are returned.
```
Could we restructure this slightly? It's a little confusing to have "if the returned object is CSV, only the 'results' are returned" directly after the discussion of the pagination headers; it could be interpreted as saying that CSV results don't get pagination. Possibly structure like so:

- Talk about pagination headers
- Talk about the result, the contents of each list item, and how it's returned directly as a list when in CSV format
- Talk about how the result list is embedded in an object with extra pagination controls when JSON is used

Up to you, though.

pomegranited replied:
Ah, thank you for the suggestion!  I think it's much clearer now.  See [63979d2](https://github.com/edx/edx-analytics-data-api/pull/128/commits/63979d2453205f4e6abee1bdd7c30e1828b9015b)

haikuginger commented:
:+1:, conditional on getting Travis to go green. 

- [x] I tested this: I set up an analytics devstack, filled it with data, and confirmed that I was able to both query it and receive CSV data in response, and was able to restrict what data was returned by specifying the `fields` URL parameter.
- [x] I read through the code
- [x] Includes documentation

pomegranited commented:
@mulby This PR is ready for your review, whenever you have time.

pomegranited commented:
@haikuginger Apologies for the change, but the client has requested that we include the current date/time in the downloaded CSV filename.

Could you please review [c7a6032](https://github.com/edx/edx-analytics-data-api/pull/128/commits/c7a6032a8271c149d24b5e2cfe90704c6e38d654)?

haikuginger commented:
:+1: on [c7a6032](https://github.com/edx/edx-analytics-data-api/pull/128/commits/c7a6032a8271c149d24b5e2cfe90704c6e38d654); tested locally and confirmed that the CSV generates a file download with the appropriate details.

bradenmacdonald commented:
Comments while testing another PR:

The default default order of fields seems to be unordered or explicitly set to "account_url, cohort, email, engagements.discussion_contributions, engagements.problem_attempts_per_completed, engagements.problems_attempted, engagements.problems_completed, engagements.videos_viewed, enrollment_date, enrollment_mode, last_updated, name, segments.0, username"

Can we get a nicer default like "username,name,email,cohort,segment,language,year_of_birth,gender,level_of_education,mailing_address,goals,city,country,first_enrollment_date,last_updated" instead?

And can there be a better name for that one field besides "segments.0" ?

bradenmacdonald commented:
Actually, a bit more information on what seems like a bug:

When I query `GET /api/v0/learners/?course_id=course-v1%3AOpenCraft%2BDemo%2B2016-08`, I get a CSV with these fields:
* account_url
* cohort
* email
* engagements.discussion_contributions
* engagements.problem_attempts_per_completed
* engagements.problems_attempted
* engagements.problems_completed
* engagements.videos_viewed
* enrollment_date
* enrollment_mode
* last_updated
* name
* segments.0
* username

But if I try to explicitly request those fields with

`GET /api/learner_analytics/v0/learners.csv?course_id=course-v1%3AOpenCraft%2BDemo%2B2016-08&fields=account_url%2Ccohort%2Cemail%2Cengagements.discussion_contributions%2Cengagements.problem_attempts_per_completed%2Cengagements.problems_attempted%2Cengagements.problems_completed%2Cengagements.videos_viewed%2Cenrollment_date%2Cenrollment_mode%2Clast_updated%2Cname%2Csegments.0%2Cusername`

Then I get a CSV with only these fields:
* account_url
* cohort
* email
* enrollment_date
* enrollment_mode
* name
* username

So it seems the `last_updated` field and any fields with a `.` cannot be included if `fields` is specified, but are included by default.

pomegranited commented:
@bradenmacdonald Thank you for your comments!

* The problem with explicitly providing a default field list is, as new fields are added, they won't be included by default (unless the default field list is also updated).
* The `segments.0` thing is because `segments` is an ordered list, and so the CSV renderer flattens it.  Would you rather I provided a single comma- or pipe-separated list instead?
* It looks like the current default list is alphabetical; I'll update the [configuration PR](https://github.com/edx/configuration/pull/3305) to reflect this.
* And yep, the issue with explicitly specifying some fields is definitely a bug.  Will address.

bradenmacdonald commented:
> The problem with explicitly providing a default field list is, as new fields are added, they won't be included by default (unless the default field list is also updated).

I see. If there's a simple way to always pull "username" and "name" to the front of the fields list without explicitly setting it, that would be nice.

> The segments.0 thing is because segments is an ordered list, and so the CSV renderer flattens it. Would you rather I provided a single comma- or pipe-separated list instead?

Yes, I think comma-separated would be nicer. That's just one opinion though.

pomegranited commented:
@bradenmacdonald Ok, I've figured out a nice way to fix the bug with the `fields` parameter not letting us specify the full fields list, and to allow the default fields order to be sensible.  It means that we'll only have fields filtering on CSV rendered responses, not JSON or other formats, but I think that's fine.

Question for you: If I specify a field that doesn't exist in the record, should there still be a column added to the CSV?  Or should that field be omitted?  e.g. if `fields=username,chicken,sandwich`, then should the CSV file return:
```
username,chicken,sandwich
user1,,
user2,,
```

bradenmacdonald commented:
@pomegranited  No, if you specify a column that doesn't exist, it shouldn't be added. Because I want to know the difference between "here are those chicken values you asked for - they're all blank" and "chicken? we don't have chicken around here. (Check for a typo in your fields list, or a version conflict, or a missing plugin.)"

bradenmacdonald commented:
BTW once you push updates, can @haikuginger please re-review? I'm pretty low on time at the moment.

haikuginger commented:
@bradenmacdonald: will do.

pomegranited commented:
@haikuginger I believe I've addressed all of Braden's points, though I wasn't able to set a nice default order; the default order is still alphabetical.

Could you please have another look?

pomegranited commented:
Rebased  [e3b1ce8](https://github.com/edx/edx-analytics-data-api/pull/128/commits/e3b1ce85f774378e80e00dd6ba361fa24d795a40) from [master](https://github.com/edx/edx-analytics-data-api/commit/fee33f4cd6da824ceae995ee60dc0d94f2c55b11) and squashed commits from to [13a0bea](https://github.com/edx/edx-analytics-data-api/pull/128/commits/13a0bea7bedd5de3677639a03a19bc946e534f0a).

haikuginger commented:
:+1:; reverified everything, and double-checked good behavior on @bradenmacdonald's cases.

thallada commented on analytics_data_api/v0/views/__init__.py:
```diff
@@ -21,3 +23,81 @@ def get(self, request, *args, **kwargs):
+        course_key = CourseKey.from_string(self.course_id)
+        course_id = u'-'.join([course_key.org, course_key.course, course_key.run])
+        now = timezone.now().replace(microsecond=0)
+        return u'{0}--{1}--{2}.csv'.format(course_id, now.isoformat(), self.filename_slug)
```
I don't think this function is unit tested. I see it called in the tests to generate the expected filename.

pomegranited replied:
@thallada Sorry, I wasn't very clear with my naming.

`v0.tests.views.test_courses` contained a method called `get_csv_filename`, which just returned a string.  [`assertCsvResponseIsValid`](https://github.com/open-craft/edx-analytics-data-api/blob/8a8edad5745e54a2e21ceadbc2d34875ddff68f3/analytics_data_api/v0/tests/views/__init__.py#L48-L54) compares that test string against the Content-disposition filename generated by the real `get_csv_filename` method.

To alleviate confusion, I've renamed the test method to a property in [8a8edad](https://github.com/edx/edx-analytics-data-api/pull/128/commits/8a8edad5745e54a2e21ceadbc2d34875ddff68f3).

thallada replied:
Ah, I see! My mistake. Thanks for renaming that!

pomegranited commented:
@thallada I've addressed your comments, rebased, and squashed in preparation for merge.  Is there anything else that needs to be fixed before this PR can be merged?  Thank you!

thallada commented:
@pomegranited Since you rebased off of your changes in #129, the tests' expected data sets need to be updated to include the new fields. Other than that, this lgtm :+1:

pomegranited commented:
Gah, sorry @thallada !  Tests fixed.
